File: tools/openburnbar-mcp/cloudvault_search_legacy.py
"""Named rollback implementation for local-MCP CloudVault search transforms."""

from __future__ import annotations

import hashlib
import hmac
import re

OPENBURNBAR_TOKEN_SEARCH_SALT = b"OpenBurnBar-CloudSearch-Salt-v1"
OPENBURNBAR_TOKEN_SEARCH_INFO = b"OpenBurnBar-CloudSearch-TokenHash-v1"
OPENBURNBAR_SEMANTIC_SEARCH_SALT = b"OpenBurnBar-CloudSearch-Semantic-Salt-v1"
OPENBURNBAR_SEMANTIC_SEARCH_INFO = b"OpenBurnBar-CloudSearch-SemanticHash-v1"
OPENBURNBAR_STOPWORDS = {
    "the",
    "and",
    "for",
    "with",
    "that",
    "this",
    "from",
    "how",
    "what",
    "where",
    "when",
    "why",
    "are",
    "was",
    "were",
    "you",
    "your",
    "have",
    "has",
    "had",
    "into",
    "onto",
    "can",
    "could",
    "should",
    "would",
}


def _hkdf_sha256(input_key: bytes, salt: bytes, info: bytes, length: int) -> bytes:
    effective_salt = salt if salt else bytes(32)
    prk = hmac.new(effective_salt, input_key, hashlib.sha256).digest()
    output = b""
    previous = b""
    counter = 1
    while len(output) < length:
        previous = hmac.new(prk, previous + info + bytes([counter]), hashlib.sha256).digest()
        output += previous
        counter += 1
    return output[:length]


def _cloud_normalized_tokens(text: str) -> list[str]:
    tokens = re.split(r"[^a-z0-9]+", text.lower())
    return [token for token in tokens if len(token) >= 2 and token not in OPENBURNBAR_STOPWORDS]


def _cloud_token_hashes(text: str, vault_key: bytes, limit: int = 10) -> list[str]:
    search_key = _hkdf_sha256(vault_key, OPENBURNBAR_TOKEN_SEARCH_SALT, OPENBURNBAR_TOKEN_SEARCH_INFO, 32)
    seen: set[str] = set()
    hashes: list[str] = []
    for token in _cloud_normalized_tokens(text):
        if token in seen:
            continue
        seen.add(token)
        hashes.append(hmac.new(search_key, token.encode("utf-8"), hashlib.sha256).digest()[:16].hex())
        if len(hashes) >= limit:
            break
    return hashes


def _simple_semantic_stem(token: str) -> str:
    suffixes = [
        "ization",
        "ations",
        "ation",
        "ments",
        "ment",
        "ingly",
        "edly",
        "ing",
        "ies",
        "ied",
        "ers",
        "er",
        "ed",
        "s",
    ]
    for suffix in suffixes:
        if len(token) > len(suffix) + 3 and token.endswith(suffix):
            stem = token[: -len(suffix)]
            return stem + "y" if suffix in {"ies", "ied"} else stem
    return token


def _cloud_semantic_features(tokens: list[str]) -> list[tuple[str, float]]:
    features: list[tuple[str, float]] = []
    seen: set[str] = set()

    def append(name: str, weight: float) -> None:
        if name and name not in seen:
            seen.add(name)
            features.append((name, weight))

    for token in tokens:
        append(f"token:{token}", 2.4)
        stem = _simple_semantic_stem(token)
        if stem != token:
            append(f"stem:{stem}", 1.8)
        if len(token) >= 5:
            append(f"prefix:{token[:5]}", 0.8)
    for index in range(0, max(0, len(tokens) - 1)):
        append(f"bigram:{tokens[index]}_{tokens[index + 1]}", 1.3)
    return features


def _cloud_semantic_hashes(text: str, vault_key: bytes, limit: int = 12) -> list[str]:
    tokens = _cloud_normalized_tokens(text)
    if not tokens or limit <= 0:
        return []
    search_key = _hkdf_sha256(vault_key, OPENBURNBAR_SEMANTIC_SEARCH_SALT, OPENBURNBAR_SEMANTIC_SEARCH_INFO, 32)
    features = _cloud_semantic_features(tokens)
    accumulator = [0.0] * 64
    for name, weight in features:
        digest = hmac.new(search_key, name.encode("utf-8"), hashlib.sha256).digest()
        index = ((digest[0] << 8) | digest[1]) % 64
        accumulator[index] += (1.0 if (digest[2] & 1) == 0 else -1.0) * weight
    hashes: list[str] = []
    seen: set[str] = set()

    def append_bucket(bucket: str) -> None:
        if len(hashes) >= limit:
            return
        digest = hmac.new(search_key, bucket.encode("utf-8"), hashlib.sha256).digest()[:16].hex()
        if digest not in seen:
            seen.add(digest)
            hashes.append(digest)

    for band in range(8):
        value = 0
        for bit in range(8):
            if accumulator[band * 8 + bit] >= 0:
                value |= 1 << bit
        append_bucket(f"simhash:v1:band:{band}:{value:02x}")
    for name, _weight in features[: max(0, limit - len(hashes))]:
        append_bucket(f"feature:v1:{name}")
    return hashes

File: tools/openburnbar-mcp-remote/src/legacy/cloudVaultSearchLegacy.ts
import { createHmac, hkdfSync } from "node:crypto";

const TOKEN_SEARCH_SALT = Buffer.from("OpenBurnBar-CloudSearch-Salt-v1");
const TOKEN_SEARCH_INFO = Buffer.from("OpenBurnBar-CloudSearch-TokenHash-v1");
const SEMANTIC_SEARCH_SALT = Buffer.from("OpenBurnBar-CloudSearch-Semantic-Salt-v1");
const SEMANTIC_SEARCH_INFO = Buffer.from("OpenBurnBar-CloudSearch-SemanticHash-v1");
const STOPWORDS = new Set([
  "the", "and", "for", "with", "that", "this", "from", "how", "what", "where", "when", "why",
  "are", "was", "were", "you", "your", "have", "has", "had", "into", "onto", "can", "could",
  "should", "would",
]);

function normalizedTokens(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/u)
    .filter((token) => token.length >= 2 && !STOPWORDS.has(token));
}

export function legacyCloudVaultTokenHashes(text: string, vaultKey: Uint8Array, limit: number): string[] {
  if (limit <= 0) {
    return [];
  }
  const searchKey = Buffer.from(hkdfSync("sha256", vaultKey, TOKEN_SEARCH_SALT, TOKEN_SEARCH_INFO, 32));
  try {
    const hashes: string[] = [];
    const seen = new Set<string>();
    for (const token of normalizedTokens(text)) {
      if (seen.has(token)) {
        continue;
      }
      seen.add(token);
      hashes.push(createHmac("sha256", searchKey).update(token).digest().subarray(0, 16).toString("hex"));
      if (hashes.length >= limit) {
        break;
      }
    }
    return hashes;
  } finally {
    searchKey.fill(0);
  }
}

function simpleSemanticStem(token: string): string {
  for (const suffix of ["ization", "ations", "ation", "ments", "ment", "ingly", "edly", "ing", "ies", "ied", "ers", "er", "ed", "s"]) {
    if (token.length > suffix.length + 3 && token.endsWith(suffix)) {
      const stem = token.slice(0, -suffix.length);
      return suffix === "ies" || suffix === "ied" ? `${stem}y` : stem;
    }
  }
  return token;
}

function semanticFeatures(tokens: string[]): Array<{ name: string; weight: number }> {
  const features: Array<{ name: string; weight: number }> = [];
  const seen = new Set<string>();
  const append = (name: string, weight: number): void => {
    if (!name || seen.has(name)) {
      return;
    }
    seen.add(name);
    features.push({ name, weight });
  };
  for (const token of tokens) {
    append(`token:${token}`, 2.4);
    const stem = simpleSemanticStem(token);
    if (stem !== token) {
      append(`stem:${stem}`, 1.8);
    }
    if (token.length >= 5) {
      append(`prefix:${token.slice(0, 5)}`, 0.8);
    }
  }
  for (let index = 0; index < tokens.length - 1; index += 1) {
    append(`bigram:${tokens[index]}_${tokens[index + 1]}`, 1.3);
  }
  return features;
}

export function legacyCloudVaultSemanticHashes(text: string, vaultKey: Uint8Array, limit: number): string[] {
  const tokens = normalizedTokens(text);
  if (tokens.length === 0 || limit <= 0) {
    return [];
  }
  const searchKey = Buffer.from(hkdfSync("sha256", vaultKey, SEMANTIC_SEARCH_SALT, SEMANTIC_SEARCH_INFO, 32));
  try {
    const accumulator = Array.from({ length: 64 }, () => 0);
    const features = semanticFeatures(tokens);
    for (const feature of features) {
      const digest = createHmac("sha256", searchKey).update(feature.name).digest();
      const index = ((digest[0] << 8) | digest[1]) % accumulator.length;
      accumulator[index] += (digest[2] & 1) === 0 ? feature.weight : -feature.weight;
    }
    const hashes: string[] = [];
    const seen = new Set<string>();
    const appendBucket = (bucket: string): void => {
      if (hashes.length >= limit) {
        return;
      }
      const hash = createHmac("sha256", searchKey).update(bucket).digest().subarray(0, 16).toString("hex");
      if (!seen.has(hash)) {
        seen.add(hash);
        hashes.push(hash);
      }
    };
    for (let band = 0; band < 8; band += 1) {
      let value = 0;
      for (let bit = 0; bit < 8; bit += 1) {
        if (accumulator[band * 8 + bit] >= 0) {
          value |= 1 << bit;
        }
      }
      appendBucket(`simhash:v1:band:${band}:${value.toString(16).padStart(2, "0")}`);
    }
    for (const feature of features.slice(0, Math.max(0, limit - hashes.length))) {
      appendBucket(`feature:v1:${feature.name}`);
    }
    return hashes;
  } finally {
    searchKey.fill(0);
  }
}
